src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -1,100 +1,215 @@
+                if (!project.SupportsCompilation || project.FilePath == null)
                 {
-                    return SpecializedTasks.EmptyTask;
+                    return;
                 }
 
-                if (_solutionId != null && _solutionId != solution.Id)
+                await UpdateReferencesAync(project, cancellationToken).ConfigureAwait(false);
+
+                var version = await project.GetSemanticVersionAsync(cancellationToken).ConfigureAwait(false);
+                Tuple<VersionStamp, SymbolTreeInfo> tuple;
+                if (_projectToInfo.TryGetValue(project.Id, out tuple) && tuple.Item1 == version)
                 {
-                    ReportCount();
-                    return SpecializedTasks.EmptyTask;
+                    return;
                 }
 
-                _solutionId = solution.Id;
-                foreach (var projectId in solution.ProjectIds)
+                var info = await SymbolTreeInfo.GetInfoForSourceAssemblyAsync(project, cancellationToken).ConfigureAwait(false);
+                tuple = Tuple.Create(version, info);
+                _projectToInfo.AddOrUpdate(project.Id, tuple, (_1, _2) => tuple);
+            }
+
+            private async Task UpdateReferencesAync(Project project, CancellationToken cancellationToken)
+            {
+                Compilation compilation = null;
+                foreach (var reference in project.MetadataReferences.OfType<PortableExecutableReference>())
                 {
-                    if (!_symbolCountByProjectMap.ContainsKey(projectId))
-                    {
-                        return SpecializedTasks.EmptyTask;
-                    }
+                    compilation = await UpdateReferenceAsync(project, reference, compilation, cancellationToken).ConfigureAwait(false);
                 }
-
-                ReportCount();
-                return SpecializedTasks.EmptyTask;
             }
 
-            public override void RemoveProject(ProjectId projectId)
+            private async Task<Compilation> UpdateReferenceAsync(
+                Project project, PortableExecutableReference reference, Compilation compilation, CancellationToken cancellationToken)
             {
-                if (_symbolCountByProjectMap != null)
+                var key = GetReferenceKey(reference);
+                if (key != null)
                 {
-                    _symbolCountByProjectMap.Remove(projectId);
+                    var lastWriteTime = GetLastWriteTime(key);
+                    if (!lastWriteTime.Item1)
+                    {
+                        // Couldn't get the write time.  Just ignore this reference.
+                        return compilation;
+                    }
+
+                    Tuple<DateTime, SymbolTreeInfo, HashSet<ProjectId>> tuple;
+                    if (_metadataPathToInfo.TryGetValue(key, out tuple) && tuple.Item1 == lastWriteTime.Item2) 
```

Comment:
Agreed.  Will Promote to real types.  Indeed, I wish we had a refactoring for that :)